server/src/main/java/org/opensearch/action/admin/cluster/remotestore/repository/RemoteStoreRepositoryRegistrationHelper.java

```diff
@@ -0,0 +1,202 @@
+        }
+
+        return settings.build();
+    }
+
+    // TODO: Add logic to mark these repository as System Repository once thats merged.
+    // Visible For testing
+    public static RepositoryMetadata buildSegmentRepositoryMetadata(DiscoveryNode node) {
+        String name = node.getAttributes().get(REMOTE_STORE_SEGMENT_REPOSITORY_NAME_ATTRIBUTE_KEY);
+        String type = node.getAttributes().get(REMOTE_STORE_SEGMENT_REPOSITORY_TYPE_ATTRIBUTE_KEY);
+        String stringSettings = node.getAttributes().get(REMOTE_STORE_SEGMENT_REPOSITORY_SETTINGS_ATTRIBUTE_KEY);
+
+        assert name != null : "Name cannot be null";
+        assert type != null : "Type cannot be null";
+        assert stringSettings != null : "Settings cannot be null";
+
+        return new RepositoryMetadata(name, type, buildSettings(stringSettings));
+
+    }
+
+    // TODO: Add logic to mark these repository as System Repository once thats merged.
+    // Visible For testing
+    public static RepositoryMetadata buildTranslogRepositoryMetadata(DiscoveryNode node) {
+        String name = node.getAttributes().get(REMOTE_STORE_TRANSLOG_REPOSITORY_NAME_ATTRIBUTE_KEY);
+        String type = node.getAttributes().get(REMOTE_STORE_TRANSLOG_REPOSITORY_TYPE_ATTRIBUTE_KEY);
+        String stringSettings = node.getAttributes().get(REMOTE_STORE_TRANSLOG_REPOSITORY_SETTINGS_ATTRIBUTE_KEY);
+
+        assert name != null : "Name cannot be null";
+        assert type != null : "Type cannot be null";
+        assert stringSettings != null : "Settings cannot be null";
+
+        return new RepositoryMetadata(name, type, buildSettings(stringSettings));
+    }
+
+    /**
+     * Validated or adds the remote store repository to cluster state if it doesn't exist.
+     * @param joiningNode
+     * @param currentState
+     * @return updated cluster state
+     */
+    public static ClusterState validateOrAddRemoteStoreRepository(DiscoveryNode joiningNode, ClusterState currentState) {
+        List<DiscoveryNode> existingNodes = new ArrayList<>(currentState.getNodes().getNodes().values());
+
+        ClusterState newState = ClusterState.builder(currentState).build();
+
+        // TODO: Mutating cluster state like this can be dangerous, this will need refactoring.
+        if (existingNodes.size() == 0) {
+            newState = updateClusterStateWithRepositoryMetadata(currentState, buildSegmentRepositoryMetadata(joiningNode));
+            newState = updateClusterStateWithRepositoryMetadata(newState, buildTranslogRepositoryMetadata(joiningNode));
+            return newState;
+        } else {
+            compareNodeAttributes(joiningNode, existingNodes.get(0));
+        }
+
+        return newState;
+    }
+
+    private static ClusterState updateClusterStateWithRepositoryMetadata(
+        ClusterState currentState,
+        RepositoryMetadata newRepositoryMetadata
+    ) {
+        RepositoriesService.validate(newRepositoryMetadata.name());
```

Comment:
I saw the reference of validate method and that is happening during the repository registration.
https://github.com/opensearch-project/OpenSearch/blob/8d153ab1cf792ff7320ab1413aae3d34e78a3f51/server/src/main/java/org/opensearch/repositories/RepositoriesService.java#L162-L163 

But I see along with this validation, we are also validating couple of other things like repository being accessible from the cluster manager and all other nodes. And only then submitting cluster state update task to actually register the repository.

However, here we are only doing static validation on name and adding this to the cluster state directly. Should we do the repository registration in the same fashion here that also checks that the repository is accessible on the cluster manager node atleast. Also, having single flow which is used at the node join and the transport action would be highly maintainable.

Reply:
Another question - if we have supplied incorrect repository settings when does the actual validation happen where we check accessibility?